New gadget: process collector

alban opened the pull request:
# New gadget: process collector

Add a new gadget that gather the list of all processes enriched with container metadata.

This is implemented with BPF iterators, a new feature of Linux 5.8. For an introduction to BPF iterators, see:
https://devconfcz2021.sched.com/event/gmK5/ebpf-iterators

process-collector can optionally use filters as prepared by the Gadget Tracer Manager in the same way as other bcc tools such as opensnoop or execsnoop:
```
apiVersion: gadget.kinvolk.io/v1alpha1
kind: Trace
metadata:
  name: process-collector
  namespace: gadget
spec:
  node: minikube
  gadget: process-collector
  filter:
    namespace: default
    labels:
      role: demo
  runMode: Manual
  outputMode: Status
```

Implementation details:
- The BPF program is compiled twice: one time with the filter and one time without. The Go program selects the right one to use depending on the command line arguments.
- To print correct json from the BPF iterator, we cannot rely solely on seq_num to know if we've printed the first entry since entries can be filtered out. We use a temporary BPF map to keep context between calls.

## How to use

```
$ kubectl apply -f  pkg/resources/samples/trace-process-collector.yaml
$ kubectl annotate -n gadget trace/process-collector gadget.kinvolk.io/operation=start
$ kubectl get -n gadget trace/process-collector -o jsonpath='{.status.output}'
[
  {
    "tgid": 921951,
    "pid": 921951,
    "comm": "sleep",
    "mntns": 4026532639,
    "container_id": "docker://d897a9298d94ce9c8d5cc8ad233936c457710358e9fda3871fa914b11bfe5d91",
    "kubernetes_namespace": "default",
    "kubernetes_pod": "privileged-pod-4m79d",
    "kubernetes_container": "privileged-pod"
  }
]
```

## Testing done

Tested on Minikube with the docker driver on Linux 5.12.

marga-kinvolk commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,106 @@
+	return 0;
+}
+
+char _license[] SEC("license") = "GPL";
```
I'm confused by the licensing. The golang files have an Apache license header. This file doesn't have a header and is licensed GPL? Why is that?

alban replied:
When I try to use another license, I get the following from the verifier:
```
117: (85) call bpf_seq_printf#126
cannot call GPL-restricted function from non-GPL compatible program
```

I will add a header.

marga-kinvolk commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,106 @@
+	__u32 seq_num = ctx->meta->seq_num;
+	struct task_struct *task = ctx->task;
+
+	if (seq_num == 0)
```
The if block below (line 83) has braces, while this one doesn't. I personally prefer to always have braces to avoid possible bugs in the future if someone were to add a line to the block and forget to add the braces, but regardless of my personal preference the style should be consistent.

alban replied:
Right, I will fix this.
The intent was to use the kernel coding style for BPF code, so no braces for single-line blocks.

marga-kinvolk commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,106 @@
+	if (seq_num == 0)
+		BPF_SEQ_PRINTF(seq, "[\n");
+
+	if (task == (void *)0) {
```
Could it ever happen that seq_num is 0 and task is NULL? If this is a possible case, then you'd print `[\n\n]\n` is that intended?... If this is not a possible case then maybe the code can be re-written as:

```
if (task == (void *)0) {
	BPF_SEQ_PRINTF(seq, "\n]\n");
	return 0;
}

if (seq_num == 0) {
	BPF_SEQ_PRINTF(seq, "[\n");
} else {
	BPF_SEQ_PRINTF(seq, ",\n");
}
```

alban replied:
> Could it ever happen that seq_num is 0 and task is NULL?

I think it's not possible for this iterator on processes: it would mean there are zero processes running on this Linux machine. If there was no processes, nobody would be able to call this iterator.

However, I think it is worth fixing this properly because:
* We might create a collector-tcp gadget (based on [bpf_iter_tcp4.c](https://github.com/torvalds/linux/blob/3dbdb38e286903ec220aaf1fb29a8d94297da246/tools/testing/selftests/bpf/progs/bpf_iter_tcp4.c) or a collector-udp gadget (based on [bpf_iter_udp4.c](https://github.com/torvalds/linux/blob/3dbdb38e286903ec220aaf1fb29a8d94297da246/tools/testing/selftests/bpf/progs/bpf_iter_udp4.c)) and in this case, it's possible to have zero tcp or zero udp sockets. If we write the correct code pattern for collector-process, then it can just work for the other gadgets.
* We could implement filters: return a subset of processes.

marga-kinvolk commented on gadget-container/gadgets/collector-process/main.go:
```diff
@@ -0,0 +1,104 @@
+
+func main() {
+	if err := increaseRlimit(); err != nil {
+		log.Fatalf("failed to increase limit memlock limit: %s", err)
```
Is the double limit here intentional?

alban replied:
Oops

alban replied:
It comes from a copy paste from somewhere else with the same typo, initially introduced by bdb80b63. I'll fix both.

marga-kinvolk commented on gadget-container/gadgets/collector-process/main.go:
```diff
@@ -0,0 +1,104 @@
+		},
+	)
+	if err != nil {
+		log.Fatalf("creating eBPF collection: %s", err)
```
These log.Fatalf calls should actually start in uppercase and include something like "Failed to" or "Error while" as this is going to be the beginning of the line, not going to get embedded into another error up the chain.

alban replied:
Done

marga-kinvolk commented on pkg/gadgettracermanager/gadgettracermanager.go:
```diff
@@ -329,6 +387,10 @@ func NewServerWithPodInformer(nodeName string) (*GadgetTracerManager, error) {
 		containerIDsByKey: make(map[string]map[string]struct{}),
 		k8sClient:         k8sClient,
 	}
+	err = g.createMap()
+	if err != nil {
```
```suggestion
	if err = g.createMap(); err != nil {
```

marga-kinvolk commented on pkg/gadgettracermanager/gadgettracermanager.go:
```diff
@@ -354,6 +416,10 @@ func NewServer(nodeName string) (*GadgetTracerManager, error) {
 		tracers:    make(map[string]tracer),
 		k8sClient:  k8sClient,
 	}
+	err = g.createMap()
+	if err != nil {
```
```suggestion
	if err = g.createMap(); err != nil {
```

marga-kinvolk commented on pkg/gadgettracermanager/gadgettracermanager.go:
```diff
@@ -302,6 +319,47 @@ func (g *GadgetTracerManager) run() {
 }
 
+func (g *GadgetTracerManager) createMap() error {
+	// Create and pin BPF map
```
Could you expand this comment to explain the purpose of the map and how it's used in the GTM?

alban replied:
I renamed the function and added documentation in the code.

marga-kinvolk reviewed (commented):
What will happen if someone tries to use IG with a kernel prior to 5.8? Will it not work at all or will just this functionality be missing?

mauriciovasquezbernal commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,112 @@
+#include <linux/bpf.h>
```
What about namign this file in `gadget-container/gadgets/collector-process/bpf/collector-process.c` instead?

alban replied:
Done.

mauriciovasquezbernal reviewed (commented):
I did a pass on the second commit. Tried it locally and worked fine. 

I still have to review the other commit and try it in Kubernetes.

alban commented on go.mod:
```diff
@@ -1,14 +1,14 @@
 
 require (
-	github.com/cilium/ebpf v0.4.0
+	github.com/cilium/ebpf v0.6.2-0.20210702091250-ca492085341e
```
Yes it should work. v0.6.2 was released 2 days ago (including https://github.com/cilium/ebpf/pull/335). It was previously not ready when I initially updated go.mod.

alban replied:
Changed to `github.com/cilium/ebpf v0.6.2` and it works fine.

alban commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,112 @@
+}__attribute__((preserve_access_index));
+
+struct container {
+	char ContainerID[256];
+	char KubernetesNamespace[256];
+	char KubernetesPod[256];
+	char KubernetesContainer[256];
```
No reasons. Fixing...

alban commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,112 @@
+} containers SEC(".maps");
+
+
+/* From: tools/lib/bpf/bpf_tracing.h */
+/*
+ * BPF_SEQ_PRINTF to wrap bpf_seq_printf to-be-printed values
+ * in a structure.
+ */
+#define BPF_SEQ_PRINTF(seq, fmt, args...)				    \
+	({								    \
+		_Pragma("GCC diagnostic push")				    \
+		_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")	    \
+		static const char ___fmt[] = fmt;			    \
+		unsigned long long ___param[] = { args };		    \
+		_Pragma("GCC diagnostic pop")				    \
+		int ___ret = bpf_seq_printf(seq, ___fmt, sizeof(___fmt),    \
+					    ___param, sizeof(___param));    \
+		___ret;							    \
+	})
```
https://github.com/libbpf/libbpf/commit/343f63e2453269965252f5503c86e83916a12aff is not yet available in released versions of libbpf. But I should add a comment.

alban replied:
Actually, it's available in the libbpf versions we are using but in a different header. So I can delete my define and use the one in libbpf.

```diff
-#include <bpf/bpf_endian.h>
+
+/* libbpf v0.4.0 introduced BPF_SEQ_PRINTF in bpf_tracing.h.
+ * In future versions, it will be in bpf_helpers.h.
+ */
+#include <bpf/bpf_tracing.h>
```

mauriciovasquezbernal replied:
I see. Btw, we're using cilium/ebpf that **doesn't** depend on libbpf but we need libbpf for the bpf headers, right? I wondering if this is the common approach on this case.

alban replied:
I think that's correct. cilium/ebpf does not compile anything like bcc does: it only takes the ELF objects. So it should not care that we use the headers from libbpf.

I see that the cilium/ebpf repository has some headers at examples/headers/bpf_helpers.h but they don't seem to contain what we need.

mauriciovasquezbernal replied:
Got it. Perhaps having a `libbpf-headers` package could help... Anyway, requiring to have libbpf installed to compile is not a big deal.

alban commented on gadget-container/gadgets/collector-process/c/ebpf.c:
```diff
@@ -0,0 +1,112 @@
+#include <linux/bpf.h>
+
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_endian.h>
```
Correct. Removed/

alban commented on gadget-container/gadgets/collector-process/main.go:
```diff
@@ -0,0 +1,103 @@
+	err = json.Unmarshal(contents, &tasks)
+	if err != nil {
+		log.Fatalf("Failed to unmarshal from BPF iterator: %s", err)
+	}
```
I was thinking of some post-processing, but it's not necessary now.

Removed.

alban commented on gadget-container/gadgets/collector-process/main.go:
```diff
@@ -0,0 +1,103 @@
+	if !ok {
+		log.Fatalf("Failed to find BPF iterator %q", BPF_ITER_NAME)
+	}
+	dumpConnIter, err := link.AttachIter(link.IterOptions{
```
Done

alban commented:
Branch updated. I also implemented a filter with `--mntnsmap` like the other bcc gadgets. The commitmsg and PR description is updated accordingly. PTAL.

alban commented:
> What will happen if someone tries to use IG with a kernel prior to 5.8? Will it not work at all or will just this functionality be missing?

I gave it a try on AKS (kernel `5.4.0-1048-azure`) and it failed with the error:

> Failed to create BPF collection: program dump_task: CO-RE relocations: no BTF for kernel version 5.4.0-1048-azure: not supported

It misses both CONFIG_DEBUG_INFO_BTF (available from Linux 5.2 but not enabled there) and the BTF iterator feature.

In that case, the gadget collector-process does not work but it does not impact other gadgets.

mauriciovasquezbernal commented on pkg/gadgettracermanager/gadget-tracer-manager-maps.h:
```diff
@@ -0,0 +1,37 @@
+#ifndef GADGET_TRACER_MANAGER_MAPS_H
+#define GADGET_TRACER_MANAGER_MAPS_H
+
+struct ns_common {
+	unsigned int inum;
+} __attribute__((preserve_access_index));
+
+struct mnt_namespace {
+	struct ns_common ns;
+}__attribute__((preserve_access_index));
```
I'm a little bit worried what it'll create conflicts with files that include other headers defining these structures. Perhaps we could avoid that definition here and letting each tools to define / import a file defining it?

alban replied:
Done :+1:

mauriciovasquezbernal commented on pkg/gadgettracermanager/gadgettracermanager.go:
```diff
@@ -302,6 +319,62 @@ func (g *GadgetTracerManager) run() {
+	if g.containersMap == nil || c.Mntns == 0 {
+		return
+	}
+	mntnsC := uint64(c.Mntns)
+	g.containersMap.Delete(mntnsC)
```
nit: `g.containersMap.Delete(uint64(c.Mntns))`

mauriciovasquezbernal reviewed (commented):
Some minor details around. 

I tried it and worked fine:
```
$ kubectl run podubuntu --image=ubuntu:latest -- sleep 1000
# inside the gadget container
$ gadgettracermanager -call add-tracer -podname podubuntu
# take printed tracer id and append to get full path of map
$ MAP=/sys/fs/bpf/gadget/mntnsset_b0441ec07b68

$  collector-process -mntnsmap $MAP
[
  {
    "tgid": 211659,
    "pid": 211659,
    "comm": "sleep",
    "mntns": 4026534603,
    "container_id": "docker://039ea855ae345c2eed33d13379411c0c5db9e031d9611f467f03cff425f3b804",
    "kubernetes_namespace": "default",
    "kubernetes_pod": "podubuntu",
    "kubernetes_container": "podubuntu"
  }
]
```

Should this be also exposed as a plugin of kubectl? (kubectl processcollector? )

alban commented on gadget-container/gadgets/collector-process/bpf/collector-process.c:
```diff
@@ -0,0 +1,112 @@
+		return 0;
+#endif
+
+	if (*counter)
+		BPF_SEQ_PRINTF(seq, ",\n");
+
+	__sync_fetch_and_add(counter, 1);
```
This is because some processes can be filtered out, so `seq_num==0` is not a good test to know if this is the first entry to be printed because maybe the first process was filtered out, and then we must not print the `,` before the second process; that would result in invalid json.

alban commented on pkg/gadgettracermanager/gadgettracermanager.go:
```diff
@@ -302,6 +319,62 @@ func (g *GadgetTracerManager) run() {
+		Type:       ebpf.Hash,
+		KeySize:    8,
+		ValueSize:  uint32(unsafe.Sizeof(container{})),
+		MaxEntries: 1024,
```
I don't know... We also have the mntns map with 10240 (10x bigger) and the cgroupId map with 128 (8x smaller). I will use a constant to use the same maximum everywhere.

alban replied:
I introduced MAX_CONTAINER_PER_NODE

alban commented on gadget-container/gadgets/collector-process/bpf/collector-process.c:
```diff
@@ -0,0 +1,112 @@
+			/* Empty list */
+			BPF_SEQ_PRINTF(seq, "[\n]\n");
+			return 0;
+		} else {
```
Done

mauriciovasquezbernal commented on pkg/gadgettracermanager/bpf-maps.h:
```diff
@@ -0,0 +1,24 @@
+	__uint(max_entries, MAX_CONTAINER_PER_NODE);
+} containers SEC(".maps");
+
+#ifdef WITH_FILTER
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, __u64);
+	__type(value, __u32);
+	__uint(max_entries, MAX_CONTAINER_PER_NODE);
+} filter SEC(".maps");
+#endif
```
I'm a bit concerned about having multiple versions of the compiled code around. I think it increases the complexity and could be confusing in some cases. Also if we have more than one "optional" map, we'd have a lot of different combinations.

I tried to use a different approach to avoid having this conditional compilation. The idea is to have a global variable in the eBPF program to enclose the access to the map, then set this to false and replace the map by a dummy one to avoid creating it when loading the program. An example of this approach is available [here](https://github.com/kinvolk/ebpf/commits/mauricio/disable-maps-idea).  Do you think it's worth to try it out or do you prefer to keep using the conditional compilation approach?

alban replied:
I think the global variable + dummy map approach increases the complexity too in the Go code because it needs to be generated. And are you sure that the verifier can remove the dead code around the map usage?

mauriciovasquezbernal replied:
Ah got it about the generated code, yeah, without that it's a mess. I agree to keep using the conditional compilation.

mauriciovasquezbernal replied:
About the verifier, yes, it does remove the dead code.

mauriciovasquezbernal commented on gadget-container/Makefile:
```diff
@@ -1,5 +1,21 @@
+CLANG_OS_FLAGS = ""
+ifeq ($(shell grep -oP '^NAME="\K\w+(?=")' /etc/os-release), Ubuntu)
+	CLANG_OS_FLAGS="-I/usr/include/$(shell uname -m)-linux-gnu"
+endif
```
Sounds good to me.

alban commented:
Now based on #211.

alban commented on pkg/gadgettracermanager/common.h:
```diff
@@ -0,0 +1,17 @@
+#ifndef GADGET_TRACER_MANAGER_COMMON_H
+#define GADGET_TRACER_MANAGER_COMMON_H
+
+#define MAX_CONTAINER_PER_NODE 1024
```
Addressed in #211.

alban commented:
Rebased on main, now there is only 1 commit. PR description and commitmsg updated.

mauriciovasquezbernal reviewed (commented):
It looks great! I tried it (again) and worked fine!

Just to minor comments, the most important is about naming. Should it be collector-process or process-collector? The later sound more natural to me.

alban commented on pkg/gadgets/collector-process/gadget.go:
```diff
@@ -0,0 +1,95 @@
+
+func (t *Trace) Start(trace *gadgetv1alpha1.Trace) {
+	output, err := tracer.RunCollector(
+		gadgets.TracePinPathFromNamespacedName(
+			types.NamespacedName{
+				Namespace: trace.ObjectMeta.Namespace,
+				Name:      trace.ObjectMeta.Name,
+			},
+		),
```
Yes. Should we keep it like that and you will clean this code after #209 gets merged, or do you want to do a PR with a2fa08f while #209 is not ready yet?

mauriciovasquezbernal replied:
I think you can just cherry-pick and merge it with this PR if you think it's good to go.

alban commented on pkg/gadgets/collector-process/gadget.go:
```diff
@@ -0,0 +1,95 @@
+	traces map[string]*Trace
+}
+
+func (f *TraceFactory) LookupOrCreate(name types.NamespacedName) gadgets.Trace {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.traces == nil {
+		f.traces = make(map[string]*Trace)
+	}
+	trace, ok := f.traces[name.String()]
+	if ok {
+		return trace
+	}
+	trace = &Trace{}
+	f.traces[name.String()] = trace
+
+	return trace
+}
+
+func (f *TraceFactory) Delete(name types.NamespacedName) error {
+	log.Infof("Deleting %s", name.String())
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	_, ok := f.traces[name.String()]
+	if !ok {
+		log.Infof("Deleting %s: does not exist", name.String())
+		return nil
+	}
+	delete(f.traces, name.String())
+	return nil
+}
```
Yes. See also https://github.com/kinvolk/inspektor-gadget/pull/208#discussion_r679141313

alban commented:
> Should it be collector-process or process-collector? The later sound more natural to me.

I agree that process-collector sounds better. I started with collector-process because I thought there will be several collector gadgets (for processes, tcp, udp, stacks...) and it's nice they have the same prefix.

Should I change everything to process-collector?

marga-kinvolk commented:
> > Should I change everything to process-collector?
> 
> I'd say yes. Let's ask for another opinion. @marga-kinvolk what do you think about the name of this gadget (process-collector vs collector-process)?

I don't have a strong opinion. "Process Collector" sounds better as a name, but if the goal is to have a collector gadget that can collect several things, as in collector-process, collector-files, collector-tcpconn, then it might make sense to keep the collector- prefix.

alban commented:
My thinking was there would be several gadgets in the collector family. But I think it's better to have a good name. I've renamed it and addressed the other comments.

mauriciovasquezbernal reviewed (approved):
LGTM!
